text/0000-light-dom.md

```diff
@@ -0,0 +1,218 @@
+![shadow spectrum](./shadow-spectrum.png?raw=true "Shadow Spectrum")
+
+### Prior Art
+
+Most of the libraries designed to support Shadow DOM also propose a Light DOM option, with a variety of Shadow DOM features (slots, scoped styles, etc.). It includes:
+
+- [StencilJS](https://stenciljs.com/docs/styling#shadow-dom-in-stencil)
+- [LitElement](https://lit-element.polymer-project.org/api/classes/_lit_element_.litelement.html#createrenderroot)
+- [MS Fast Element](https://fast.design/docs/fast-element/working-with-shadow-dom#shadow-dom-configuration)
+
+## Detailed design
+
+### Selecting Light DOM vs Shadow DOM
+
+In raw JS, developer has the choice to use or not use shadow DOM for their custom element. For example,
+
+```js
+// uses Shadow DOM
+class MyShadowElement extends HTMLElement {
+  connectedCallback() {
+    this.attachShadow({ mode: "open" });
+    this.shadowRoot.innerHTML = "<p>My custom element that uses Shadow DOM";
+  }
+}
+
+// doesn't use Shadow DOM
+class MyLightElement extends HTMLElement {
+  connectedCallback() {
+    this.innerHTML = "<p>My element that does not use Shadow DOM";
+  }
+}
+```
+
+In LWC as well, the developer will retain the choice: they can choose to inherit from `LightningElement` to select Shadow DOM (which already exists) or inherit from a new class, `MacroElement`, to opt into Light DOM.
+
+```js
+import { MacroElement } from "lwc";
+
+export default class MyLightComponent extends MacroElement {
+  renderedCallback() {
+    this.querySelector("child-element"); // note the absence of this.template
+  }
+}
+```
+
+`MacroElement` is a new class that will be exported from the `lwc` module. Its API and functionality remain identical to `LightningElement` with the primary difference of not using Shadow DOM.
+
+### Component features when using Light DOM
+
+Some of the LWC component capabilities are directly inherited from Shadow DOM, or emulated by the synthetic-shadow. Despite the use of Light DOM, we’d like to keep these features available, even if their behavior is slightly adapted to the Light DOM:
+
+#### Slots
+
+As we mentioned before, the component composition model in LWC is provided by slots. Light DOM will provide the same mental model for developers building Light DOM components.
+
+In Light DOM, `<slot>` will denote the place where the slotted component will be attached. The `<slot>` element itself won't be rendered. The slotted content (or the fallback content) will be flattened to the parent element at runtime.
+
+Since the `<slot>` element itself isn't rendered, adding attributes or event listeners to the `<slot>` element in the template will throw a compiler error.
```

Comment:
I may be misremembering, but I thought light DOM wouldn't have slots? Should this section be removed? (Or replaced with a stub saying "If you want slots, use shadow DOM"? 😃 )

Reply:
Based on discussion, it looks like this needs to be updated with the new `macro:slot` proposal. Other questions that are worth answering in this section:

What is the timing model of slots? For instance:

```html
<foo-component>
<macro:slot></macro:slot>
<bar-component>
```

In what order do the connectedCallbacks for `foo-component`, elements within the slot, and `bar-component` get called?

Related to that question: will light DOM slots be [lazy or eager](https://dev.to/richharris/why-i-don-t-use-web-components-2cia)? My preference is for lazy, for the reasons outlined in the linked post, e.g.:

```html
<template if:true={foo}>
  <macro:slot></macro:slot>  
</template>
```

In the above case, rendering the slot is wasted effort if `foo` evaluates to `false`. Lazy slots would allow us to skip that wasted work.

Also: how will slots impact SSR? I assume SSR will just render the slotted content directly, i.e. `<macro:slot>` is never actually represented in DOM that the browser sees? But then I assume there will need to be some HTML comments so that hydration knows where the slots are. This can probably be handled in SSR RFCs though.

Reply:
Another question about timing in slots. Based on my reading of https://github.com/WICG/webcomponents/issues/909#issuecomment-743269886, there is an issue that arises with e.g.:

```html
<!-- child-component.html -->
<div>hello</div>
<macro:slot></macro:slot>
<macro:slot name="title"></macro:slot>
<div>world</div>
```

```html
<!-- parent-component.html -->
<child-component>
  <h1 macro:slot="title"></h1>
  <div>yolo</div>
</child-component>
```

There are two options here: 1. the user's content renders first, or 2) the `child-component`'s content renders first.

In either case, the two content sources could overwrite each other, e.g. by code in the `connectedCallback` or `renderedCallback` appending elements to `this`, or by reading `element.nextSibling`, etc. Because there is no shadow boundary, the two DOM sources can observe each other.

Also, note that the `title` slot and the bare slot have a different ordering in the user's code versus the `child-component`'s code. So in the `renderedCallback` for each, what element order is observed?

In that comment, the author suggests that the only way to make this work is to render the user's content into its own little isolated DOM and then "project" it into the other part of the DOM. But clearly this would involve MutationObserver or some other heavyweight mechanism to mirror the two. It definitely seems complicated.

Prior art worth investigating: [Svelte slots](https://svelte.dev/docs#slot) and [Vue slots](https://vuejs.org/v2/guide/components-slots.html).

Reply:
FWIW I tried to reproduce the performance problems with light-DOM slots described in [this post](https://medium.com/8451/make-your-stenciljs-web-components-faster-by-using-shadow-dom-d010a9f0cdda), but I can't find a repro.

Stencil's light DOM slot implementation seems to not suffer from significant costs from "moving" the DOM nodes around, so maybe there isn't really a concern here.

Reply:
Updated RFC with some clarifications on how slots will work.